Lib/typing.py

```diff
@@ -331,6 +331,12 @@ def _eval_type(t, globalns, localns, recursive_guard=frozenset()):
     if isinstance(t, ForwardRef):
         return t._evaluate(globalns, localns, recursive_guard)
     if isinstance(t, (_GenericAlias, GenericAlias, types.UnionType)):
+        if isinstance(t, GenericAlias):
+            args = tuple(
+                ForwardRef(arg) if isinstance(arg, str) else arg
```

Comment:
Do we need to pass `globals` and `locals` to `ForwardRef` here? 🤔

Reply:
I may not be knowledgable enough about the typing code base, but I don't see a globals/locals argument for `ForwardRef`. Do you mean the module? I couldn't find out how to use/what to pass to that argument exactly.

Reply:
It's too late for that anyway. The ForwardRef class has an optional module= argument which may be used to resolve references, but the best we could do at this point is passing `globalns.get("__name__")`, which would just end up with the same globalns as we are using anyway.

This suggests there could still be scenarios where this will fail, esp. when a type alias defined in one module is used in another. But that would only be solvable by recording the module at the time the alias is being defined, and we've already said we wouldn't go that far (since it would require a ForwardRef implementation in C).